File: AI/Assignment 4/OptimalSearch.py
import heapq

# Graph definition (adjacency list with costs)
graph = {
    "Pune": [("Expressway", 50), ("OldHighway", 30), ("CoastalRoad", 20)],
    "Expressway": [("Mumbai", 100)],
    "OldHighway": [("Mumbai", 120)],
    "CoastalRoad": [("Mumbai", 180)],
    "Mumbai": []
}

# Heuristic values (estimated cost to reach goal)
heuristic = {
    "Pune": 150,
    "Expressway": 100,
    "OldHighway": 120,
    "CoastalRoad": 180,
    "Mumbai": 0
}

# -------------------- A* Implementation --------------------
def a_star(start, goal):
    # Priority queue (min-heap)
    open_list = []
    heapq.heappush(open_list, (heuristic[start], 0, start))  # (f, g, node)

    g_score = {start: 0}
    parent = {start: None}

    while open_list:
        f, g, current = heapq.heappop(open_list)

        # Goal check
        if current == goal:
            path = []
            while current:
                path.append(current)
                current = parent[current]
            path.reverse()
            print("Path found:", " -> ".join(path))
            print("Total cost:", g)
            return

        # Explore neighbours
        for neighbor, cost in graph[current]:
            tentative_g = g + cost
            if neighbor not in g_score or tentative_g < g_score[neighbor]:
                g_score[neighbor] = tentative_g
                parent[neighbor] = current
                f_new = tentative_g + heuristic[neighbor]
                heapq.heappush(open_list, (f_new, tentative_g, neighbor))

    print("No path found!")


# -------------------- MAIN --------------------
if __name__ == "__main__":
    print("Running A* Search from Pune to Mumbai...\n")
    a_star("Pune", "Mumbai")

File: AI/Assignment 4/OptimalSearch.cpp
// A* Search Algorithm Example
#include <bits/stdc++.h>
using namespace std;

// Structure for edges (directed, with weight)
struct Edge {
    string to;
    int cost;
};

// Heuristic values (h(n)) – estimated cost to reach goal
unordered_map<string, int> heuristic = {
    {"Pune", 150},
    {"Expressway", 100},
    {"OldHighway", 120},
    {"CoastalRoad", 180},
    {"Mumbai", 0}
};

// Graph definition (Adjacency list)
unordered_map<string, vector<Edge>> graph = {
    {"Pune", {{"Expressway", 50}, {"OldHighway", 30}, {"CoastalRoad", 20}}},
    {"Expressway", {{"Mumbai", 100}}},
    {"OldHighway", {{"Mumbai", 120}}},
    {"CoastalRoad", {{"Mumbai", 180}}},
    {"Mumbai", {}}
};

// Node structure for priority queue
struct Node {
    string name;
    int g; // cost so far
    int f; // g + h
    bool operator>(const Node& other) const {
        return f > other.f;
    }
};

// -------------------- A* Implementation --------------------
void aStarSearch(string start, string goal) {
    priority_queue<Node, vector<Node>, greater<Node>> open;
    unordered_map<string, int> gScore;
    unordered_map<string, string> parent;

    gScore[start] = 0;
    open.push({start, 0, heuristic[start]});

    while (!open.empty()) {
        Node current = open.top();
        open.pop();

        // Goal reached
        if (current.name == goal) {
            cout << "Path found: ";
            vector<string> path;
            for (string at = goal; !at.empty(); at = parent[at]) {
                path.push_back(at);
            }
            reverse(path.begin(), path.end());
            for (auto& city : path) cout << city << " ";
            cout << "\nTotal cost: " << current.g << endl;
            return;
        }

        // Explore neighbours
        for (auto& edge : graph[current.name]) {
            int tentative_g = current.g + edge.cost;
            if (!gScore.count(edge.to) || tentative_g < gScore[edge.to]) {
                gScore[edge.to] = tentative_g;
                parent[edge.to] = current.name;
                int f = tentative_g + heuristic[edge.to];
                open.push({edge.to, tentative_g, f});
            }
        }
    }

    cout << "No path found!" << endl;
}

// -------------------- MAIN --------------------
int main() { 
    cout << "Running A* Search from Pune to Mumbai...\n";
    aStarSearch("Pune", "Mumbai");
    return 0;
}
